### PR TITLE
Deserialize the KV

### PR DESCRIPTION
**Pull request by ashamis**
Part of #1299

### CODE REVIEW

File: src/ds/champ_map.h [hunk 1]
```diff
@@ -345,16 +345,17 @@ namespace champ
       while (size != 0)
       {
         // Deserialize the key
-        auto key_size = serialized::read<uint64_t>(data, size);
-        const uint8_t* key = data;
-        serialized::skip(data, size, key_size + get_padding(key_size));
+        uint64_t key_size = size;
```
Comment (jumaffre) [inline src/ds/champ_map.h:348]: `size_t`

Comment (jumaffre) [reply src/ds/champ_map.h:348]: Also, this seems inconsistent with the type of `key_size` in `get_buffer()`.

File: src/kv/store.h [hunk 2]
```diff
@@ -219,35 +219,76 @@ namespace kv
...
+    }
 
+    std::unique_ptr<Snapshot> snapshot(Version v) override
+    {
+      auto snapshot = std::make_unique<Snapshot>(v);
+
+      {
+        std::lock_guard<SpinLock> mguard(maps_lock);
+
+        if (v > current_version())
```
Comment (jumaffre) [inline src/kv/store.h:268]: We also need to check that `v >= commit_version()`.

File: src/kv/store.h [hunk 3]
```diff
@@ -219,35 +219,76 @@ namespace kv
...
     {
-      std::lock_guard<SpinLock> mguard(maps_lock);
-
-      if (v > current_version())
       {
-        throw ccf::ccf_logic_error(fmt::format(
-          "Attempting to snapshot at invalid version v:{}, current_version:{}",
-          v,
-          current_version()));
+        std::lock_guard<SpinLock> vguard(version_lock);
```
Comment (jumaffre) [inline src/kv/store.h:225]: It looks like these changes should be applied _after_ the snapshot has been applied.

File: src/kv/untyped_map.h [hunk 4]
```diff
@@ -682,7 +690,41 @@ namespace kv::untyped
...
+
+      std::function<SerialisedEntry(const uint8_t*& key, size_t&)> make_k =
+        [](const uint8_t*& data, size_t& size) -> SerialisedEntry {
+        uint64_t key_size = serialized::read<uint64_t>(data, size);
+        SerialisedEntry ret;
+        ret.assign(key_size, *data);
+        serialized::skip(data, size, key_size);
+        return std::move(ret);
+      };
+      std::function<kv::VersionV<SerialisedEntry>(const uint8_t*& key, size_t&)>
```
Comment (jumaffre) [inline src/kv/untyped_map.h:712]: `kv::VersionV<SerialisedEntry>` ->`VersionV`

Comment (jumaffre) [reply src/kv/untyped_map.h:712]: Same in `snapshot()`.

File: src/kv/untyped_map.h [hunk 5]
```diff
@@ -682,7 +690,41 @@ namespace kv::untyped
...
+        roll.commits->get_head() == roll.commits->get_tail(),
+        "We are apply a snapshot and there are pending commits");
+
+      std::function<SerialisedEntry(const uint8_t*& key, size_t&)> make_k =
+        [](const uint8_t*& data, size_t& size) -> SerialisedEntry {
+        uint64_t key_size = serialized::read<uint64_t>(data, size);
+        SerialisedEntry ret;
+        ret.assign(key_size, *data);
+        serialized::skip(data, size, key_size);
+        return std::move(ret);
```
Comment (jumaffre) [inline src/kv/untyped_map.h:710]: No need for `std::move()` here?

File: src/kv/store.h (hunk 3 above)
Comment (ashamis) [reply src/kv/store.h:225]: I am simplifying this but the order should not matter. We are skipping forward an arbitrary number of version so there should not be any other transactions in progress.

File: src/kv/store.h (hunk 2 above)
Comment (ashamis) [reply src/kv/store.h:268]: why >= ? Should we not be able to create a snapshot of the current version?

Comment (jumaffre) [reply src/kv/store.h:266]: `commit_version()` refers to the version at which the map was compacted (i.e. the first element in the roll). We should not be able to snapshot before this version.

**Pull request merged**